lib/ueberauth/strategy/cognito.ex

```diff
@@ -36,14 +38,27 @@ defmodule Ueberauth.Strategy.Cognito do
       client_id: client_id
     } = Config.get_config(otp_app(conn))
 
-    params = %{
-      response_type: "code",
-      client_id: client_id,
-      redirect_uri: callback_url(conn),
-      state: state,
-      # TODO - make dynamic (accepting PRs!):
-      scope: "openid profile email"
-    }
+    optional_params = @accepted_authorize_params
+    |> Enum.map(fn key ->
+      case Map.fetch(conn.params, "#{key}") do
+        {:ok, param} -> {key, param}
+        _ -> nil
+      end
+    end)
+    |> Enum.filter(& &1)
+    |> Map.new()
```

Comment:
This can be a `flat_map` instead of a `map` + `filter`:
```elixir
@accepted_authorize_params
|> Enum.flat_map(fn key ->
  case Map.fetch(conn.params, Atom.to_string(key)) do
    {:ok, value} -> [key: value]
    _ -> nil
  end
end)
|> Map.new()
```

Reply:
That's certainly the best way to go.

Thank you.

Reply:
I'm sorry.
I've tried.
Cannot return `nil` with `flat_map`.
I was wondering if it could exclude the `nil`...

Reply:
```elixir
iex(2)> [1, 2, 3] |> Enum.flat_map(fn x -> if x > 1, do: [], else: [x: 100] end)
[x: 100]
```
It worked.

Reply:
Yep, that's what I get for writing code in the GH comment boxes :)

Reply:
```elixir
iex(2)> [:hoge, 100, :foo, 200] |> Map.new
** (ArgumentError) argument error
    (stdlib 3.12) :maps.from_list([:hoge, 100, :foo, 200])
    (elixir 1.10.2) lib/map.ex:181: Map.new/1
```

Is there any better way to get this going?
```elixir
iex(2)> [{:hoge, 100}, {:foo, 200}] |> Map.new
%{foo: 200, hoge: 100}
```

Reply:
I'm sorry.
This worked.

```elixir
    optional_params = @accepted_authorize_params
    |> Enum.flat_map(fn key ->
      case Map.fetch(conn.params, Atom.to_string(key)) do
        {:ok, value} -> [{key, value}]
        _ -> []
      end
    end)
    |> Map.new()

```

Reply:
The second version is the same as the Keyword list:

```elixir
iex(1)> [a: 1, b: 2]
[a: 1, b: 2]
iex(2)> [{:a, 1}, {:b, 2}]
[a: 1, b: 2]
```
If you return a (possibly empty) Keyword list from the `flat_map` function, it should go into `Map.new/1`.

```elixir
enum
|> Enum.flat_map(fn key ->
  if valid? do
    [{:key, 1}] # Keyword list
  else
    [] # empty Keyword list
  end
end)
|> Map.new()
```